### PR TITLE
Allow end user to select a choice different from list of available choices in profile_list

### PR DESCRIPTION
**Pull request by batpad**
Opening a Draft PR working toward #699 .

This should eventually:

 - Allow specifying an option on profiles to allow a users to specify free-form text for an image (currently a single boolean called `allow_other`, but I think @yuvipanda had some ideas to make that a bit nicer.
 - If `allow_other` is true, show an additional option called "Other" in the dropdown and handle showing a text input when Other is selected
 - Handle the text input value on the backend to actually pull the image onto the cluster, etc.

Currently, have added a simple `allow_other` option and handling it just on the frontend.

@yuvipanda hopefully this is a good starting point for us to work together on implementing handling the value on the backend and improve the semantics of the options.

cc @consideRatio

### CODE REVIEW

**Comment by consideRatio** [comment]
Ooooh great to see you work towards this @batpad!!!

### Reflection on config naming

I think `allow_other` could be renamed to be a bit more self-explanatory, as something related to the `choices` config.

With a free-form input field, front-end validation of the input is probably going to end up being another feature. I don't think we need to think about implementing that as part of this PR, but I'm considering it already when thinking on what configuration to provide. For example `include_other_choice` as a boolean may need to be augmented with `other_choice_validation` or similar, which makes me consider if we want `other_choice` as a dictionary with `included` under it or similar, allowing `other_choice.included` and `other_choice.some_validation_related_config`.

Hmmm hmmm, I'm currently leaning towards a flat structure to configure this, like `other_choice_included`, to be complemented with `other_choice_some_validation_related_config`. There are pros and cons of using a flat config structure (`a_b`, `a_c`) vs a nested structure (`a["b"]`, `a["c"]`), but I think staying flat is fine here but its good if we can find a common prefix in the name such as `other_choice_`.

Wdyt about renaming `allow_other` (boolean) to `other_choice_included` (boolean)?

**Comment by batpad** [comment]
@consideRatio I think @yuvipanda had a good idea for how this config should be structured, making it an object, that also contained a regex to use to validate the user input. Let's wait for him to weigh in on the structure for options - sorry, he had pasted that to me somewhere, but I can't find it :-/ .

**Comment by batpad** [comment]
> There are pros and cons of using a flat config structure (a_b, a_c) vs a nested structure (a["b"], a["c"]), but I think staying flat is fine here but its good if we can find a common prefix in the name such as other_choice_.

Hmm, this is an interesting one - yea, I do think this should allow for multiple "sub-options" like the regex to validate input, so either prefixing with `other_choice_` or making it a nested structure sounds good to me - I don't really have a preference on nested vs flat - happy to go with whatever @yuvipanda and you feel is best here.

**Comment by yuvipanda** [comment]
Thanks for your comments, @consideRatio!

I left a comment in https://github.com/batpad/kubespawner/commit/cc0037b9753d6905cd1d98a0a2acaded905fd84d#r111675499 earlier, which I will repro here.

Let's expand this a little bit. How about something like:

```yaml
freeform:
   enabled: true
   match_regex: (optional regex to match)
   kubespawner_override:
       some_config_key: some_value-with-{value}-substituted-with-what-user-wrote
```

So for an image where we only want to allow pangeo images, it would be something like

```yaml
freeform:
   enabled: true
   match_regex: ^pangeo/.*$
   kubespawner_override:
      image: "{value}"
```

https://github.com/jupyterhub/kubespawner/blob/ff8dcf15205208fb25a7255563b59f53f07aca88/kubespawner/spawner.py#L3016 is what gets the form data, converts them into 'user options' that are then passed on to https://github.com/jupyterhub/kubespawner/blob/ff8dcf15205208fb25a7255563b59f53f07aca88/kubespawner/spawner.py#L3091, which interprets them and applies the overrides.

**Comment by yuvipanda** [comment]
With respect to nested vs flat, @consideRatio, I think nested is better here! Because we already will have three things:

1. Wether to enable this or not
2. How it should actually override the kubespawner traitlet config. Currently, with choices, we set a 'key' and override it with with specific config that is set in the `jupyterhub_config.py`. With free-form text, we should allow admins the option to interpolate the value entered into an override. We also need to *store* the raw user entered value in user_options, and re-setup the kubespawner_override whenever the user server is stopped / started again.
3. Simple regex based frontend validation, that @batpad tells me can be done with HTML5 form validation built into browsers, no extras needed.

Given this, I'd suggest the nested config structure.

**Comment by batpad** [comment]
To add:

 - We would also want something like a `validation_message` to show a human-readable error message / describe the input format required, like the human readable value of the `regex` field.

**Comment by yuvipanda** [comment]
Re: nested config, I also opened https://github.com/jupyterhub/kubespawner/issues/736 to help us mitigate the complexity of validation in the nested config in the future. Does not need to block this PR though. I would like us to make sure we get frontend validation in here with this though.

File: jupyterhub_config.py
Comment (consideRatio) [inline jupyterhub_config.py:49]: @yuvipanda @batpad are you okay naming this `other_choice`?

I think the "other choice" name provides intuition that:
- its one additional choice to `choices`
- of a special "free input" kind
- that its just one single choice addition

I don't think naming this `free_form` would provide as much intuition on what the configuration is about compared to `other_choice`. And if not, it would making this config harder to understand intuitively, which could make it harder to and setup and update over time.

Comment (batpad) [reply jupyterhub_config.py:49]: > @yuvipanda @batpad are you okay naming this other_choice?

`other_choice` sounds great to me :-)

Will wait a bit to see if @yuvipanda has any strong opinions, but I can go ahead and make that change.

Do the other variable names inside the object sound okay? Maybe be nice to make all naming changes in one go.

Thanks again for the thoughtfulness around this.

Comment (consideRatio) [reply jupyterhub_config.py:49]: - `enabled` :+1:
- `display_name` :+1: - consistent with other naming
- `match_regex` :thinking: --- works, but I can also see value in having `validation_` be involved in this naming, if it is `validation_match_regex` or `validation_regex`
- `validation_message` :+1: 
- `kubespawner_override` :+1: - consistent with other naming

Comment (batpad) [reply jupyterhub_config.py]: Nice!

@consideRatio I have made a commit renaming `free_form` to `other_choice` and `match_regex` to `validation_match_regex`.

> match_regex 🤔 --- works, but I can also see value in having validation_ be involved in this naming, if it is validation_match_regex or validation_regex. I think I favor validation_match_regex.

I definitely like prefixing with `validation` since we also have `validation_message` and also STRONGLY agree that `validation` should not be further nested. Personally, I think I'd prefer the slightly less verbose `validation_regex` to `validation_match_regex` as `validation_regex` sounds pretty unambiguous to me. But also not a hard opinion, and I'm not as familiar with general naming conventions here, so went with your suggestion here. If you do agree / feel that `validation_regex` is fine, am happy to change that.

Thank you again.

Comment (batpad) [reply jupyterhub_config.py]: Ah, forgot to mention - also changed the `name` for the input field that's submitted with the POST request from `profile-option-{{ profile.slug }}-{{ k }}-other` to `profile-option-{{ profile.slug }}-{{ k }}-other-choice` to keep that consistent with the name of the option.

@consideRatio would be great to get your 👀 and 👍 / 👎 on the naming of this as well, since that's the interface between the front and backend and it would be nice to be thoughtful about what we call this.

**Comment by ranchodeluxe** [comment]
Hey folks, just checking in here. There's a demo on June 26th that Aimee will be doing and I'm wondering if maybe this work could be wrapped up and exposed on `nasa-veda` cluster by then? No worries if it's a stretch. Is there anyway I can help?

**Comment by batpad** [comment]
Starting to look at the backend integration for this.

Just pasting over @yuvipanda's comments from elsewhere about where this needs to happen to have it easy to refer to:

https://github.com/jupyterhub/kubespawner/blob/ff8dcf15205208fb25a7255563b59f53f07aca88/kubespawner/spawner.py#L3016 is what gets the form data, converts them into 'user options' that are then passed on to https://github.com/jupyterhub/kubespawner/blob/ff8dcf15205208fb25a7255563b59f53f07aca88/kubespawner/spawner.py#L3091, which interprets them and applies the overrides.

I'll start familiarizing myself here and then lean on you @yuvipanda to get through the actual implementation.

**Comment by batpad** [comment]
@yuvipanda @consideRatio there's probably some discussion to be had around details of the implementation here, but this is now "complete" and ready for review:

 - Adds an `other_choice` option which handles showing a free text input for that choice
 - Handles validating and overriding the appropriate profile value on the backend
 - Has basic tests for the above, including a test to handle the `choices` parameter which did not have a test

Overall, I think there's some ideas on how we could do this whole thing a bit better, but that might be a bit of a larger refactor, that hopefully we can discuss separately.

I'll give this whole thing another good look tomorrow, but do let me know anything that pops out / that should definitely be fixed.

(thanks again @yuvipanda for helping me figure out most of the backend integration code)

File: jupyterhub_config.py
Comment (consideRatio) [reply jupyterhub_config.py]: > Personally, I think I'd prefer the slightly less verbose `validation_regex` to `validation_match_regex` as `validation_regex` sounds pretty unambiguous to me.

I think you are right that it sounds pretty unambiguous and I couldn't figure out how to misunderstand it!

> If you do agree / feel that `validation_regex` is fine, am happy to change that.

Thank you for thinking about naming things so thoroughly!!! Let's go for it!

**Comment by consideRatio** [comment]
> * Currently, the value for the "other" option in the drop-down is hard-coded as "other". This means you can't have a different option in the drop-down that has a value "other". Hopefully this is fine, but just a heads-up - we can change this hard-coded "other" to something else if this seems like it might be a problem.

I think its the right call to hardcode this specific choice for a given profile, and "other" seems like a fine choice but also one that could risk collide with other choices. I think either we should make the name a bit more obscure, like `other-choice` / `other_choice`, or force the other choices slugs to not overlap with the help of validation logic.

File: jupyterhub_config.py
Comment (batpad) [reply jupyterhub_config.py]: > Thank you for thinking about naming things so thoroughly @batpad!!! Let's go for it!

Made this change :-)

**Comment by batpad** [comment]
>  I think either we should make the name a bit more obscure, like other-choice / other_choice, or force the other choices slugs to not overlap with the help of validation logic.

Speaking to @yuvipanda a bit here -- have gone with naming the form field with `--other-choice` i.e. adding an additional `-` to prevent accidental naming conflicts.

It would also be good to have comprehensive validation of the config via JSON schema or similar, and then we can just prohibit `choices` with `--` in them. We can do that as a separate PR?

**Comment by batpad** [comment]
> I think either we should make the name a bit more obscure, like other-choice / other_choice, or force the other choices slugs to not overlap with the help of validation logic.

Apologies, I mis-spoke in my comment above. Two different things needed to be done:

 - Change the value of the `name` field, affecting the key for the value being POSTed, made the `--other-choice` instead of `-other-choice`
 - Changing the `value` in the drop-down for the `Other` option. Changed that currently to just `other-choice` from `other`.

File: kubespawner/spawner.py
Comment (yuvipanda) [inline kubespawner/spawner.py:1548]: Specify that you can use `{value}` here to substitute whatever the user entered value was.

File: kubespawner/spawner.py
Comment (yuvipanda) [inline kubespawner/spawner.py:3074]:
```suggestion
                    # other_choice is enabled
```

File: kubespawner/spawner.py
Comment (yuvipanda) [inline kubespawner/spawner.py:3077]: Given `{option_name}--other-choice` is used multiple times, can we make it into a variable for clarity? `other_choice_form_key` or something?

File: kubespawner/spawner.py
Comment (yuvipanda) [inline kubespawner/spawner.py:3094]: You can directly call `re.match` here instead. If you look at the cpython source code, it gets compiled automatically (https://github.com/python/cpython/blob/904aef994262383ae916545908f0578c2d53cf31/Lib/re/__init__.py#L163). And compile also caches it (https://github.com/python/cpython/blob/904aef994262383ae916545908f0578c2d53cf31/Lib/re/__init__.py#L273) so it isn't recompiled unless necessary. Just having a .match here is cleaner than a compile and match.

File: kubespawner/templates/style.css
Comment (yuvipanda) [inline kubespawner/templates/style.css:31]: bootstrap 3, which is used here, has a .hidden that we can use instead https://getbootstrap.com/docs/3.4/css/#helper-classes

File: tests/test_spawner.py
Comment (yuvipanda) [inline tests/test_spawner.py:961]: A docstring describing what this is testing would be great!

File: tests/test_spawner.py
Comment (yuvipanda) [inline tests/test_spawner.py:980]: A docstring here would be great too

File: tests/test_spawner.py
Comment (yuvipanda) [inline tests/test_spawner.py:999]: Docstring <3

File: tests/test_spawner.py
Comment (yuvipanda) [inline tests/test_spawner.py:1019]: Thank you for adding all these tests! A docstring would be great here too.

**Review by yuvipanda — CHANGES_REQUESTED** [review]
Some minor feedback, but otherwise this LGTM!

File: kubespawner/spawner.py
Comment (batpad) [reply kubespawner/spawner.py]: I agree.

The problem here is that we still need to define the variable in two places, since we have these two different loops. So, I had to create the variable twice: https://github.com/jupyterhub/kubespawner/pull/735/commits/f55e32302a9e80a967c9531fbf10068ca6d7970f#diff-6fdb3a928956e8ab2b810679d51fae72870d45ed4a088e85e16814339aff8901R3074 and https://github.com/jupyterhub/kubespawner/pull/735/commits/f55e32302a9e80a967c9531fbf10068ca6d7970f#diff-6fdb3a928956e8ab2b810679d51fae72870d45ed4a088e85e16814339aff8901R3107 .

This feels really not nice. I think options:

 - Make get_other_choice_key a function or so and call the function from both places so at least we define the `--other-choice` constant in only one place.
 - Refactor to not use two loops and loop only once
 - Leave as is in my latest commit
 - Something else .. ?

Thanks much for the review!

**Comment by echarles** [comment]
How will this be used in the z2jh helm chart? Looking at https://z2jh.jupyter.org/en/stable/resources/reference.html#singleuser-profilelist I don't see a direct/obvious relation/impact, so wanted to ask.

**Comment by yuvipanda** [comment]
@echarles once this is merged, released, and bumped in z2jh, someone would probably add a documentation PR there so it can be used.

File: jupyterhub_config.py
Comment (yuvipanda) [inline jupyterhub_config.py:49]: `other_choice` always seemed a bit *too* generic to me, as `other` feels super generic a term. I spent some more time thinking about this, and I think I'd prefer calling it `unlisted_choice`, as it makes it clearer there is a *list* of choices from the admin, and this enables the end user to make a choice that is *unlisted*, but still possibly supported with some validation.

**Review by yuvipanda — COMMENTED** [review]
This looks good to me, but I'd love for the name of the config to be different. I've made a suggestion, but otherwise this all seems fine.

Thanks for working through this, @batpad

**Comment by echarles** [comment]
> @echarles once this is merged, released, and bumped in z2jh, someone would probably add a documentation PR there so it can be used.

Maeks sense, I have opened https://github.com/jupyterhub/zero-to-jupyterhub-k8s/issues/3152 to follow that.

File: jupyterhub_config.py
Comment (consideRatio) [reply jupyterhub_config.py:49]: I'm not sure what I think is the best naming yet. I agree there is benefit of understanding that `profile_options` can involve a dropdown list of choices, but it seems a bit out of scope to let this name, which may or may not show up in a config, help convey that we have a list of choices.

I realize I've had a hidden assumption: that this new feature can be used without entries in the `choices` list. I think this is related to the naming discussion because it influences how coupled this choice is to the `choices` list.

The new config in this PR can be used when no entries are in `choices`, it just comes with some redundant UI:

### `choices` with _two_ entries, `other_choice` enabled

current

### `choices` with _no_ entries, `other_choice` enabled

current-with-no-listed-choices

---

@yuvipanda and @batpad here are some decisions I think would be helpful to make before deciding on the naming:

1. If the new config `other_choice`/`unlisted_choice` be allowed when `choices` is empty
2. What UI to go for in those situations
3. If it should say "Other..." or something else by default in the dropdown list when using `other_choice`/`unlisted_choice`
4. If that "Other..." text should be configurable or not

I think (1) it would be great to support use of the new config when `choices` is empty, and (2) that it would be great to not show a single-value dropdown list in the case that the new config is enabled with `choices` empty. I'm not sure about (3) and (4).

Comment (yuvipanda) [reply jupyterhub_config.py:49]: @consideRatio I agree with you on 1 and 2, and should probably be configurable as well (4). I think (Other) is a reasonable default when we do have a list of choices. However, I think none of these should block this PR getting merged, and can be done as follow up PRs.

I think this being usable without any listed choices and the ability to configure the “other” text makes the case stronger for calling it “unlisted_choice” than “other_choice”.

**Comment by consideRatio** [comment]
Hey @batpad I'm going on vacation for three weeks! Below are two documentation related review points, but besides those lean on other maintainers guidance - I'd love to see this approved and merged! Thank you for your work on this @batpad, I think it will be a very appreciated feature!!!

1. Is the new config `help` string documenting the functionality well enough via the configuration reference and reflecting the state to be merged?
   - Reviewable via [PR preview of the built documentation](https://jupyterhub-kubespawner--735.org.readthedocs.build/en/735/spawner.html#kubespawner.KubeSpawner.profile_list)
   - It could be relevant to involve `.. versionchanged:: 6.1` to clarify new functionality was added in the `profile_list` config was added at that point in time.
   - Note that the content of the `help` string is rST formatted, not MyST markdown.
   - Note that documentation can be built and viewed locally with `pip install -r docs/requirements.txt`, `cd docs`, `make devenv`. I'm not sure if it auto-refreshes correctly and re-generates built documentation from the source code, maybe. Otherwise do `rm -rf _build` and run again.
2. Is the PR title and description reflecting the final state to be merged?

File: kubespawner/spawner.py
Comment (GeorgianaElena) [reply kubespawner/spawner.py:3073]: Actually, isn't `selected_profile_user_options` what's passed over by the user in the UI through the form (or though API) and `profile.get('profile_options')` all the options available to them?

If that's the case, then shouldn't we check that the selected options are valid options and not the other way around? Hmm, I'm a bit confused about this now.

cc @yuvipanda

**Review by GeorgianaElena — COMMENTED** [review]
Thanks a lot for the PR @batpad and thanks for adding tests <3

I left a comment/question about the validation done in the backend, but that's mostly me making sure I understood the implementation. This generally looks good 👍🏼 Thank you ✨

File: jupyterhub_config.py
Comment (GeorgianaElena) [reply jupyterhub_config.py:49]: If it makes any difference, I think `unlisted_choice` it's more accurate with respect to what it represents.
But I don't think strongly about naming, as long as we have docs that do a good job describing what it does.

File: kubespawner/spawner.py
Comment (yuvipanda) [reply kubespawner/spawner.py:3073]: @GeorgianaElena By having them in different loops, we don't do any processing at all of the options if *any* of the options provided were invalid. We could split them up into different functions, but I think they should probably just be two loops.

Comment (yuvipanda) [reply kubespawner/spawner.py:3073]: > If that's the case, then shouldn't we check that the selected options are valid options and not the other way around? Hmm, I'm a bit confused about this now.

So the behavior change of this is that options that aren't configured by the admin but sent over will be ignored, while previously they will be errored. Is that right? I do think (and suggested!) the new code, as I think it's much clearer than what was there before. Do you think ignoring extra values while making sure whatever the admin has sent is validated is problematic?

File: jupyterhub_config.py
Comment (yuvipanda) [reply jupyterhub_config.py:49]: I'd love for us to go with `unlisted_choice`!

**Comment by yuvipanda** [comment]
Unfortunately, @batpad is a little sick, so I'm taking this on and trying to push it forward to completion now.

File: kubespawner/spawner.py
Comment (yuvipanda) [reply kubespawner/spawner.py]: I've moved the validation to a different function and added a docstring

Comment (yuvipanda) [reply kubespawner/spawner.py]: @GeorgianaElena I've adapted your PR into https://github.com/batpad/kubespawner/pull/2, what do you think?

**Review by GeorgianaElena — APPROVED** [review]
I believed this should be merged, after https://github.com/batpad/kubespawner/pull/2 gets merged as well! 🚀

**Review by yuvipanda — APPROVED** [review]

**Comment by yuvipanda** [comment]
@GeorgianaElena ok this is ready to go now i think!

After this, I think it'd be useful to do a refactor to move all the profile handling code into a separate file. But I don't want to make this PR too much longer...

**Comment by GeorgianaElena** [comment]
@yuvipanda, I agree! But for now, let's merge this! 🚀 

Many thanks to everyone for all the thought and effort that went into this PR! 🎉

File: jupyterhub_config.py
Comment (yuvipanda) [reply jupyterhub_config.py]: I opened https://github.com/jupyterhub/kubespawner/issues/757 and https://github.com/jupyterhub/kubespawner/issues/756 to track these.

**Pull request merged**
